sdks/python/src/lilypad/_internal/otel/_utils.py

```diff
@@ -0,0 +1,77 @@
+"""Utility classes and functions for Lilypad OpenTelemetry instrumentation."""
+
+import base64
+import logging
+from typing import Any, Protocol, Sequence, cast, runtime_checkable
+
+from opentelemetry.util.types import AttributeValue
+
+from ..utils import json_dumps
+
+logger = logging.getLogger(__name__)
+
+
+@runtime_checkable
+class InstrumentedClient(Protocol):
+    """[MISSING DOCSTRING]"""
+
+    __lilypad_instrumented_client__: bool
+
+
+def client_is_already_instrumented(client: Any) -> bool:
+    """[MISSING DOCSTRING]"""
+    if (
+        isinstance(client, InstrumentedClient)
+        and client.__lilypad_instrumented_client__
+    ):
+        logger.debug(f"{type(client).__name__} has already been instrumented, skipping")
+        return True
+    return False
+
+
+def mark_client_as_instrumented(client: Any) -> None:
+    """[MISSING DOCSTRING]"""
+    cast(InstrumentedClient, client).__lilypad_instrumented_client__ = True
+
+
+def serialize_attribute_value(value: Any) -> AttributeValue | None:
```

Comment:
nit: i'd have added unit tests for this function (eg for cases like the mixed float int case)